### PR TITLE
Add object_async example

### PR DESCRIPTION
**Pull request by GretaCB**
This object-async example builds off the async standalone example and the object-sync example

### Next Actions
- [x] Add code comments and relevant docs ("[When would you use an async object or class?](https://github.com/mapbox/node-cpp-skel/blob/85c81269be34bf28d50c92e8ffe98b4901d699a2/README.md#when-would-you-use-an-object-or-class)")
- [x] All possible Test Coverage for this example
- [x] Add some context around `node::ObjectWrap` and how it relates to `Unwrap`

### Questions
- @springmeyer Does this example hit any of the questions we pondered in https://github.com/mapbox/node-cpp-skel/issues/50 ?
- Does this custom constructor design make sense with an async object? 
- Is this PR relevant to real world usage of an async object?
- Curious exactly what is `node::ObjectWrap` and how it relates to `Unwrap`

Any other thoughts/comments/edits?

cc @springmeyer @daniel-j-h @flippmoke

### CODE REVIEW

File: src/object_async/hello_async.cpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  *
+  * v8::Local<v8::Function> callback;
+  * CallbackError("error message", callback);
+  * return; // this is important to prevent duplicate callbacks from being fired!
+  */
+  inline void CallbackError(std::string message, v8::Local<v8::Function> callback) {
+  	v8::Local<v8::Value> argv[1] = { Nan::Error(message.c_str()) };
+  	Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+  }
+
```
Comment (springmeyer) [inline src/object_async/hello_async.cpp]: I see this same function getting repeated in multiple `.cpp` files. How about we refactor the code to move it to a new header file called `module_utils.hpp`? Then define it there, include `module_utils.hpp` in each of the `cpp` files that uses it, and that will reduce the code duplication.

Note: for fun context because of the `inline` keyword is used the compiler is doing away with this function in the binary code. When something is "inlined" it is re-arranged and optimized. So the `v8::Local<v8::Value>...` and `Nan::MakeCallback...` lines are being moved to where the code is calling `CallbackError`. This means that moving the source code into a common header will only impact how the source code looks. The final binary will probably not change at all because of how the compiler works.

File: src/object_async/hello_async.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  };  
+
+
+  NAN_METHOD(HelloObjectAsync::hello_async) {
+    // "info" comes from the NAN_METHOD macro, which returns differently
+    // according to the version of node
+    // Mention anything about "Unwrap"?
+    HelloObjectAsync* h = Nan::ObjectWrap::Unwrap<HelloObjectAsync>(info.Holder());
+
+    std::string name = h->name_;
```
Comment (springmeyer) [inline src/object_async/hello_async.cpp]: Minor issue: this line is resulting in an uncessary copy. We are copying the `h->name_` to a new variable named `name`. Instead, down below you could do:

```
new AsyncHelloWorker{louder, h->name_, new Nan::Callback{callback}};
```

That would avoid the copy. Or alternatively you could do:

```c++
std::string const& name = h->name_;
```

That would also avoid the copy by making the `name` variable a [reference](https://github.com/mapbox/cpp/blob/master/glossary.md#reference) to the `h->name_`.

Why does this matter?

In short, performance. If this code needed to be as fast as possible we want to avoid all possible copies. Because copies require allocation (or re-allocation of memory) for the new variables space. And `std::string` allocates memory on the heap (internally) (https://github.com/mapbox/cpp/blob/master/glossary.md#heap-allocation) when newly created.

File: src/object_async/hello_async.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  */
+  inline void CallbackError(std::string message, v8::Local<v8::Function> callback) {
+  	v8::Local<v8::Value> argv[1] = { Nan::Error(message.c_str()) };
+  	Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+  }
+
+
+  // Expensive allocation of std::map, querying, and string comparison,
+  // therefore threads are busy
+  std::string do_expensive_work(bool louder, std::string name) {
```
Comment (springmeyer) [inline src/object_async/hello_async.cpp]: To avoid a copy here pass by reference like:

```c++
std::string do_expensive_work(bool louder, std::string const& name) {  
```

File: src/object_async/hello_async.cpp [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+  }  
+
+  // This is the worker running asynchronously and calling a user-provided callback when done.
+  // Consider storing all C++ objects you need by value or by shared_ptr to keep them alive until done.
+  // Nan AsyncWorker docs: https://github.com/nodejs/nan/blob/master/doc/asyncworker.md
+  struct AsyncHelloWorker : Nan::AsyncWorker {
+
+      using Base = Nan::AsyncWorker;  
+
+      AsyncHelloWorker(bool louder, std::string name, Nan::Callback* callback)
```
Comment (springmeyer) [inline src/object_async/hello_async.cpp]: Another spot where we can avoid a copy:

Best to pass by reference like:

```c++
AsyncHelloWorker(bool louder, std::string const& name, Nan::Callback* callback)
```

File: src/object_async/hello_async.cpp [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+              else
+              {
+                return Nan::ThrowTypeError(
+                    "arg must be a string");
+              }
+            }
+            else {
+                auto *const self = new HelloObjectAsync();
+                self->Wrap(info.This());
+            }
```
Comment (springmeyer) [inline src/object_async/hello_async.cpp]: I think a more realistic thing would be to throw an error if no argument is passed. Currently we default to `world` below. But I think instead we should throw a JS error here if no `name` arg is passed. What do you think?

**Comment by springmeyer** [comment]
> @springmeyer Does this example hit any of the questions we pondered in #50 ?

Yes, perfectly. We are passing in a custom class that holds a custom data member. That member is safe to pass into the threadpool because it is not a V8 object, but rather a `std::string`.

> Does this custom constructor design make sense with an async object?

You mean the custom constructor that accepts the `name` arg? Yes, I think this makes sense. However I did make one inline comment above that I think we should require this argument rather than allowing the object to be created without an arg.

> Is this PR relevant to real world usage of an async object?

It is. Well done!

> Curious exactly what is node::ObjectWrap and how it relates to Unwrap

We spoke about this voice. The short version is that `node::ObjectWrap` and wrapping/unwrapping objects is the somewhat clumsy way it is possible to bind the two languages. Let's not dig deeper than just knowing:

 - To access a class instance inside a C++ `static method` you must unwrap the object.
 - The C++ methods must be static to make them available at startup across the language boundary.

File: src/object_async/hello_async.cpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            return Nan::ThrowTypeError(ex.what());
+        }
+
+        info.GetReturnValue().Set(info.This());
+    }
+    else
+    {
+        return Nan::ThrowTypeError(
+            "Cannot call constructor as function, you need to use 'new' keyword");
+    }
```
Comment (daniel-j-h) [inline src/object_async/hello_async.cpp]: I think we could handle both `new T()` as well as `T()` by adding this to the function's beginning:

```c++
if (!info.IsConstructCall()) {
  auto init = Nan::New(create_once());
  info.GetReturnValue().Set(init->NewInstance());
  return;
}

/*continue with New function impl*/
```

Not sure if we want to do that, though.

If not we could flatten the control flow here by exiting early on `!IsConstructorCall()`.

File: src/object_async/hello_async.cpp (hunk 2 above)
Comment (daniel-j-h) [inline src/object_async/hello_async.cpp]: Since you're pointing out string copies here - not sure if it makes sense to use move-semantics throughout this example (e.g. to move the string into the async worker). Maybe it just makes it more complicated than it has to be..

Comment (springmeyer) [inline src/object_async/hello_async.cpp]: 👍  great suggestion @daniel-j-h - I agree that move semantics here would likely make the most sense given that the goal of node-cpp-skel is to set developers up for writing high performance code.

File: src/object_async/hello_async.cpp (hunk 6 above)
Comment (springmeyer) [inline src/object_async/hello_async.cpp]: @daniel-j-h - really great catch 🙇 . I agree that handling the case where a user forgets to use `new` in Javascript is important.  I'm aware that some JS developers like to hide the complexity and allow object creation to work without `new`. But my preference would be to throw since this is a consistent pattern that has worked in other node C++ modules I have written. And I've never seen a complaint about this.

File: src/object_async/hello_async.cpp (hunk 1 above)
Comment (GretaCB) [inline src/object_async/hello_async.cpp]: @springmeyer so `inline` is great for reusable methods within a file?

Comment (springmeyer) [inline src/object_async/hello_async.cpp]: > @springmeyer so inline is great for reusable methods within a file?

- inlining, when in a cpp file, may help performance.
- inlining, when in an hpp file, may help performance and may also be absolutely necessary to allow the header to be included from multiple cpp files when the code in the header includes both the definition and the implementation (which a "header-only" library does). And example of this being fixed is https://github.com/mapbox/vector-tile/pull/19.

So, inlining is important to understand because it impacts performance, code size, and is critical to header-only library design.

> And a util header is one step further than that, great for reusable methods within an application/directory

Exactly. If you want to share code in multiple .cpp files then you need to put at least definitions for it in an .hpp file (and, in the case of header-only design, also the implementation)

Ticketed adding this to the C++ glossary at https://github.com/mapbox/cpp/issues/28

Comment (springmeyer) [inline src/object_async/hello_async.cpp]: Finished at https://github.com/mapbox/cpp/pull/29

File: src/object_async/hello_async.cpp (hunk 6 above)
Comment (GretaCB) [inline src/object_async/hello_async.cpp]: Going to lean towards being more explicit here and requiring `new`

**Comment by GretaCB** [comment]
Thank you so much for your input @daniel-j-h @springmeyer 🙌 

I've updated the PR with the optimization suggestions and comments/links for context. Once tests are 🍏  , I will merge! Feel free to add any other comments if I missed anything.
